### ISSUE TITLE
Implement plugins non-blocking

### ISSUE THREAD

**Issue by agaida** [lxqt/lxqt-panel#1064]
##### Expected Behavior

Starting a plugin should not cause the terminl to hang 25s
Esp. true for statusnotifier plugin. 

##### Current Behavior

Panel is blocked in some cases and appears 25s after desktop start

##### Possible Solution

Fix it? - Hint might be: https://smcv.pseudorandom.co.uk/2008/11/nonblocking/
https://forum.lxqt.org/t/lxqt-panel-appearing-with-huge-delay-when-home-directory-moved/715/51

##### Steps to Reproduce (for bugs)

1. Find a installation with this bug - in my case debian testing when i plugi in my logitech cam
2. Restart

System: Debian testing and latest git and others.

**Comment by tsujan** [comment]
More info:

I only saw it in Debian Buster on VirtualBox. Removal of Status Notifier "fixed" the long delay at login. But adding it and replacing xfwm4 with openbox also "fixed" it.

While the long delay happened at login, it didn't happen when the panel was started from terminal or LXQt Session Settings.

It seems that different users see it under different circumstances — a change that "fixes" it for one doesn't work for another — but there's no delay *without* Status Notifier.

**Comment by palinek** [comment]
Please, can you provide a suitable backtrace of the blocked lxqt-panel process?

**Comment by agaida** [comment]
Will try - but it is hard, because no crashes and systemstart. Btw, i ported back your time out patch - now the delay is down to approx 5s - a value i could live with. But it also applies here: We have to wait, until the patch hit testing and new iso's with the patch are built.

**Comment by tsujan** [comment]
> We have to wait, until the patch hit testing and new iso's with the patch are built.

Tell me when it's ready. One more test on VirtualBox won't do harm.

> the delay is down to approx 5s

Which mean no delay, IMO.

**Comment by tsujan** [comment]
Never mind! I enabled xfwm4 again and installed the latest git libqtxdg on Buster's ISO — `sudo make install` is safe inside a virtual machine. I tested it 4 times, with reboot too. The 30-sec delay is gone. Panel shows up after 5 sec and is usable after 4 more sec. I don't know whether I can call it "a short delay" or "no delay".

**Comment by tsujan** [comment]
Booted the virtual machine again; it's still 5 + 4 (or 5) sec. The second 4(or 5)-sec delay is definitely in Status Notifier but this is a virtual machine, after all.

**Comment by palinek** [comment]
Can you, guys, give the #1065 a chance?

**Comment by tsujan** [comment]
> Can you, guys, give the #1065 a chance?

I'm trying it -- will take a while; everything is much slower on virtual machines.

**Comment by tsujan** [comment]
https://github.com/lxqt/lxqt-panel/pull/1065 had no effect on the remaining short delay of Status Notifier but, as I said before, I don't know whether a 4-6-sec delay counts on a virtual machine.

**Comment by agaida** [comment]
some of the 4-5 sec might be related to xfconfig. @tsujan - you took libqtxdg from buildd-unstable?

**Comment by tsujan** [comment]
> @tsujan - you took libqtxdg from buildd-unstable?

I installed it directly from git -- otherwise, I would see the 30-sec delay with xfwm4, which was fixed after git installation.

Status Notifier does have a delay on VirtualBox and, IMHO, we need to find its cause but it's only 4-6 sec here.

**Comment by agaida** [comment]
ok, encountered the same - tbh, i can live with these 4s, i can't with 25-30s - great. Let's test a day or two and merge.

**Comment by tsujan** [comment]
> Let's test a day or two and merge.

As I said above, https://github.com/lxqt/lxqt-panel/pull/1065 did NOT have effect here -- actually, it may have increased the delay by 1-2 sec. What HAD effect was @palinek's previous patch (https://github.com/lxqt/libqtxdg/commit/7a0a262b4b9b8ec0ba5420b9eb1667c641083474), which is already in git.

**Comment by agaida** [comment]
we had to test it on more faulty installations - please keep in mind that we don't even know about the problems.

**Comment by tsujan** [comment]
> We had to test it on more faulty installations

Right.

My summary:

https://github.com/lxqt/libqtxdg/commit/7a0a262b4b9b8ec0ba5420b9eb1667c641083474 does the job and, *if there's still a short delay* (about which, I'm not sure), https://github.com/lxqt/lxqt-panel/pull/1065 does not remove it (but may increase it by 1-2 sec). Just that.

Since I've never seen a delay on my real system, I can't test anything else. But, the more diverse are the tests, the better.

**Comment by palinek** [comment]
> lxqt/libqtxdg@7a0a262 does the job 

If this fixes the delayed startup then the cause of the problem should be some autostarted `DBusActivatable` application...so the blocking should be in lxqt-session. AFAIK lxqt-panel doesn't start any apps on its startup.

**Comment by palinek** [comment]
> #1065 did NOT have effect here -- actually, it may have increased the delay by 1-2 sec.

Hm...I doubt that. If it even would need to spawn a new thread for the concurrent execution, it wouldn't take so much time...

**Comment by tsujan** [comment]
> Hm...I doubt that. If it even...

I haven't read the code — just tested it last night — and, since all tests were done with VirtualBox, I can't be sure either (as I mentioned above).

**Comment by palinek** [comment]
Just a note... I've just tried https://cdimage.debian.org/images/unofficial/non-free/images-including-firmware/buster_di_rc1-live+nonfree/amd64/iso-hybrid/debian-live-buster-DI-rc1-amd64-lxqt+nonfree.iso in virtual machine and seeing no delay after DM login

**Comment by palinek** [comment]
@agaida Is this still a problem?

**Comment by agaida** [comment]
good question :)

**Comment by 1ras** [comment]
This still affects Debian Bullseye which is the next stable release of Debian.

**Comment by stefonarch** [comment]
I'm running bullseye with git sources and openbox for some time now (on an old 32bit laptop) and never noticed this delay. There has to be some factor involved.

**Comment by 1ras** [comment]
You are right, it looks like only xfwm4 is affected which is unfortunately the default on Debian Buster and Bullseye :-(

**Comment by stefonarch** [comment]
So there are 2 alternatives at least. I never liked xfwm4 mainly because of shortcut conflicts.

**Comment by slidinghotdog** [comment]
I didn't notice any delay/blocking here, using xfwm4 and arch for more than 1 year

**Comment by 1ras** [comment]
The shortcut notification error that pops up on every login is another annoying issue. Both issues happen out of the box when LXQt is installed with the Debian installer of Buster and Bullseye. This is certainly not how a desktop environment should be delivered.

I first saw this issues two years ago when Buster was released, so I got the impression that LXQt is not ready yet for daily use and  decided not to use it. That's why I'm so surprised that two years later, Bullseye is close to release, but the issues are still the same.

Of course that's a Debian issue. If I open a Debian bug for it, what's the upstream recommendation how it should be solved?

**Comment by tsujan** [comment]
@1ras
That should be reported to Debian, not here. There's no shortcut notification error with a standard LXQt installation.

**Comment by 1ras** [comment]
@tsujan
Sure, but what's the recommendation to fix it? E.g. replace xfwm4 with openbox? What window manager is used with a "standard LXQt installation" ?

**Comment by tsujan** [comment]
I don't know what the issue is. Debian's LXQt is outdated but the problem can't be about that because, I started LXQt when I had Debian and I didn't see any notification error since day 1. I haven't seen any with LXQt+Manjaro either. I use my packages that I compile from the git sources (I did so in Debian too).

**Comment by tsujan** [comment]
> What window manager is used with a "standard LXQt installation" ?

Any DE-agnostic WM. LXQt doesn't have a default WM. I've used Openbox, xfwm4 and KWin.

**Comment by stefonarch** [comment]
As I'm running bullseye with git I made a test, installing xfwm4 and using it with a fresh user. The popup appears saying "shortcut Ctrl+Alt+D for showing desktop cannot be registered" and in fact it's default in xfwm4. 
Of course this would be any distro's job, changing one of them, but this eventually will happen to everybody who installs lxqt and xfwm4 on any distro.

We could improve the text maybe, adding "check your window manager keyboard shortcut settings" or using Alt-D as default.

Will test also with openbox if a similar issue exists as the default config is used now.

**Comment by 1ras** [comment]
@stefonarch
Both conflicting shortcuts are for the identical function (show desktop), but they are provided by different parts of the desktop environment. When xfwm4 is in use, maybe it's an option to hide the message or disable the shortcut within lxqt-panel. Because why should a user care about two identical functions?

Openbox does not show such an issue to me, that's how I realized it's related to xfwm4.

> "You never get a second chance to make a first impression" ;)

;-)

**Comment by tsujan** [comment]
> ... but they are provided by different parts of the desktop environment.

No. The DE is LXQt and it doesn't have conflicting shortcuts. `xfwm4` is a file manager that can be used with LXQt (and is a good option) but it's the distro's job to prevent conflicts.

When @agaida was the maintainer of deb packages, Debian's LXQt was among the best. Unfortunately, he disappeared from both LXQt and Debian.

**Comment by tsujan** [comment]
> We could improve the text maybe, adding "check your window manager keyboard shortcut settings" or using Alt-D as default.

The WM isn't the only possible cause and any other shortcut may be taken by a WM.

I don't think this is LXQt's problem. KDE fixed it by not starting `kglobalaccel5` with KWin under LXQt (→ https://github.com/lxqt/lxqt/discussions/2002).

**Comment by tsujan** [comment]
> For my understanding a window manager is a part of a desktop environment

And xfwm4 isn't the only available WM.

> How?

They know their job, as @agaida did.

**Comment by tsujan** [comment]
From start, this wasn't caused by panel but by xfwm4 — probably, by a setting in it. Hence closing.
